File: main/188-20191020.go
package main

func maxProfit(k int, prices []int) int {
	if len(prices) == 0 || k == 0 {
		return 0
	}
	minInt := -1 << 32
	profits := make([][][]int, len(prices))
	for ii := 0; ii < len(prices); ii++ {
		arr := make([][]int, k+1)
		for kk := 0; kk <= k; kk++ {
			arr[kk] = []int{minInt, minInt}
		}
		profits[ii] = arr
	}

	profits[0][0] = []int{0, -prices[0]}
	size := len(prices)
	for ii := 1; ii < size; ii++ {
		for kk := 0; kk <= k; kk++ {
			if kk == 0 {
				profits[ii][0][0] = profits[ii-1][0][0]
				profits[ii][0][1] = max(profits[ii-1][0][1], profits[ii-1][0][0]-prices[ii])
			} else {
				profits[ii][kk][0] = max(profits[ii-1][kk][0], profits[ii-1][kk-1][1]+prices[ii])
				profits[ii][kk][1] = max(profits[ii-1][kk][1], profits[ii-1][kk][0]-prices[ii])
			}
		}
	}

	max := -1
	for _, arr := range profits[size-1] {
		if max < arr[0] {
			max = arr[0]
		}
	}
	return max
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// func main() {
// 	// prices := []int{2, 4, 1}
// 	prices := []int{3, 2, 6, 5, 0, 3}
// 	k := 2
// 	fmt.Println(maxProfit(k, prices))
// }

File: main/188-20191020.rb
# @param {Integer} k
# @param {Integer[]} prices
# @return {Integer}
def max_profit(k, prices)
  return 0 if prices.empty? || k.zero?
  min_int = -1<<32
  profits = []
  prices.size.times do |i|
    profits[i] = (k+1).times.map { [min_int, min_int] }
  end

  profits[0][0] = [0, -prices[0]]

  for i in (1...prices.size) do
    for kk in (0..k) do
      if kk == 0
        profits[i][0][0] = profits[i-1][0][0]
        profits[i][0][1] = [ profits[i-1][0][1], profits[i-1][0][0] - prices[i] ].max
      else
        profits[i][kk][0] = [ profits[i-1][kk][0], profits[i-1][kk-1][1] + prices[i] ].max
        profits[i][kk][1] = [ profits[i-1][kk][1], profits[i-1][kk][0] - prices[i] ].max
      end
    end
  end

  profits[-1].map(&:first).max
end

puts max_profit(2, [2,4,1])
puts max_profit(2, [3,2,6,5,0,3])